lib/install/management/dispatcher.go

```diff
@@ -211,41 +211,87 @@ func (d *Dispatcher) CollectDiagnosticLogs() {
+	pool.AppendCertsFromPEM(cas)
 
 	// update target to use FQDN
-	names, err := net.LookupAddr(ip)
-	if err != nil {
-		log.Debugf("Unable to perform reverse lookup of IP address: %s", err)
+	for _, ip := range candidateIPs {
+		names, err := net.LookupAddr(ip.String())
+		if err != nil {
+			log.Debugf("Unable to perform reverse lookup of IP address: %s", err)
+		}
+
+		// check all the returned names, and lastly the raw IP
+		for _, n := range append(names, ip.String()) {
+			opts := x509.VerifyOptions{
+				Roots:   pool,
+				DNSName: n,
+			}
+
+			_, err := cert.Verify(opts)
+			if err == nil {
+				// this identifier will work
+				log.Debugf("Matched %s for use against host certificate", n)
+				// trim '.' fqdn suffix if fqdn
+				return strings.TrimSuffix(n, "."), nil
+			}
+
+			log.Debugf("Checked %s, no match for host cert", n)
+		}
 	}
 
-	cert, err := conf.HostCertificate.X509Certificate()
-	if err != nil {
-		log.Debugf("Unable to extract host certificate: %s", err)
-		return ip, err
+	// no viable address
+	return "", errors.New("unable to determine viable address")
+}
+
+/// viableHostAddresses attempts to determine which possibles addresses in the certificate
+// are viable from the current location.
+// This will return all IP addresses - it attempts to validate DNS names via resolution.
+// This does NOT check connectivity
+func viableHostAddress(candidateIPs []net.IP, cert *x509.Certificate, cas []byte) (string, error) {
+	if cert == nil {
+		return "", fmt.Errorf("unable to determine suitable address with nil certificate")
 	}
 
-	// check all the returned names, and lastly the raw IP
-	for _, n := range append(names, ip) {
-		opts := x509.VerifyOptions{
-			Roots:   pool,
-			DNSName: n,
-		}
+	log.Debug("Loading CAs for client auth")
+	pool := x509.NewCertPool()
+	pool.AppendCertsFromPEM(cas)
+
+	dnsnames := cert.DNSNames
+
+	// assemble the common name and alt names
+	ip := net.ParseIP(cert.Subject.CommonName)
+	if ip != nil {
+		candidateIPs = append(candidateIPs, ip)
+	} else {
+		// assume it's dns
+		dnsnames = append([]string{cert.Subject.CommonName}, dnsnames...)
+	}
+
+	// turn the DNS names into IPs
+	for _, n := range dnsnames {
+		// see which resolve from here
+		ips, _ := net.LookupIP(n)
+		if len(ips) == 0 {
+			log.Debugf("Discarding name from viable set: %s", n)
+			continue
 
-		_, err := cert.Verify(opts)
-		if err == nil {
-			// this identifier will work
-			log.Debugf("Matched %s for use against host certificate", n)
-			// trim '.' fqdn suffix if fqdn
-			return strings.TrimSuffix(n, "."), nil
 		}
 
-		log.Debugf("Checked %s, no match for host cert", n)
+		candidateIPs = append(candidateIPs, ips...)
 	}
 
-	// no viable address
-	return ip, errors.New("unable to determine viable address")
+	// always add all the altname IPs - we're not checking for connectivity
+	candidateIPs = append(candidateIPs, cert.IPAddresses...)
+
+	return addrToUse(candidateIPs, cert, cas)
```

Comment:
So this resolves the ips back to dns names, while we did the reverse in this function. Why?